### PR TITLE
Reworked interpolation and added LinearInterpolator.

### PR DESCRIPTION
**Pull request by MeetThePatel**

### CODE REVIEW

File: src/autodiff/overloading/add.rs [hunk 1]
```diff
@@ -146,9 +149,9 @@ mod test_overload {
         let z = x + y;
         let grad = z.accumulate();
 
-        assert_eq!(z.value, 3.0);
-        assert_eq!(grad.wrt(&x), 1.0);
-        assert_eq!(grad.wrt(&y), 1.0);
+        assert_approx_equal!(z.value, 3.0, EPS);
```
Comment (avhz) [inline src/autodiff/overloading/add.rs:152]: `assert_approx_equal!` is not necessary in these instances I think.

File: src/autodiff/overloading/div.rs [hunk 2]
```diff
@@ -143,9 +145,9 @@ mod test_overload {
         let z = x / y;
         let grad = z.accumulate();
 
-        assert_eq!(z.value, 0.5);
-        assert_eq!(grad.wrt(&x), 0.5);
-        assert_eq!(grad.wrt(&y), -0.25);
+        assert_approx_equal!(z.value, 0.5, EPS);
```
Comment (avhz) [inline src/autodiff/overloading/div.rs:148]: As above.

File: src/autodiff/overloading/sub.rs [hunk 3]
```diff
@@ -140,9 +140,9 @@ mod test_overload {
         let z = x - y;
         let grad = z.accumulate();
 
-        assert_eq!(z.value, -1.0);
-        assert_eq!(grad.wrt(&x), 1.0);
-        assert_eq!(grad.wrt(&y), -1.0);
+        assert!((z.value - -1.0).abs() < f64::EPSILON);
```
Comment (avhz) [inline src/autodiff/overloading/sub.rs:143]: `assert_approx_equal` ?

File: src/autodiff/variables/variable.rs [hunk 4]
```diff
@@ -49,8 +49,9 @@ pub struct Variable<'v> {
 
 impl<'v> Variable<'v> {
     /// Instantiate a new variable.
+    #[must_use]
     #[inline]
-    pub fn new(graph: &'v Graph, index: usize, value: f64) -> Self {
+    pub const fn new(graph: &'v Graph, index: usize, value: f64) -> Self {
```
Comment (avhz) [inline src/autodiff/variables/variable.rs:54]: What's the reasoning here ?

File: src/autodiff/variables/variable.rs [hunk 5]
```diff
@@ -227,7 +241,7 @@ mod tests_variable {
             index: 5,
             value: std::f64::consts::PI,
         };
-        assert_eq!(var.graph() as *const _, &graph as *const _);
+        assert_eq!(var.graph() as *const _, std::ptr::addr_of!(graph));
```
Comment (avhz) [inline src/autodiff/variables/variable.rs:244]: What's the benefit ?

File: src/autodiff/variables/variable.rs (hunk 4 above)
Comment (MeetThePatel) [reply src/autodiff/variables/variable.rs:54]: It allows you to create a const variable with the constructor.

**Comment by avhz** [comment]
Hi @MeetThePatel, I'm trying to add an exponential interpolator by basically copying what you did for the linear interpolator, but due to the type/trait constraints I have been unsuccessful. 

Do you know of a solution ? The code in question is [here](https://github.com/avhz/RustQuant/blob/d6333b2275f2712ed984930a9942bbdd1030e0f4/src/math/interpolation.rs#L291)

Basically I need to use `powf` but am not able to due to the `InterpolationValue` and `InterpolationIndex` constraints.

**Pull request merged**